### PR TITLE
refactor SOCKS header decoding and add UDP request decoding

### PR DESCRIPTION
**Pull request by trevj**
This adds support for decoding UDP headers. These are very similar to "regular" SOCKS request headers so I refactored our existing code a little. Also defined TypeScript interfaces for SOCKS requests and UDP requests. I think it's a little better overall.

Next step is probably to have the UDP relay actually relay data.

### CODE REVIEW

File: spec/socks-to-rtc/socks-headers_spec.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
+describe("socks", function() {
```
Comment (iislucas) [inline spec/socks-to-rtc/socks-headers_spec.js:2]: Nice! Maybe add TODOs/issues to add ipv6 requests also.

File: src/socks-to-rtc/socks.ts [hunk 2]
```diff
@@ -325,17 +173,8 @@ module Socks {
     // Given a data |buffer|, interpret the SOCKS request.
     public static interpretRequest = (buffer:ArrayBuffer) => {
```
Comment (iislucas) [inline src/socks-to-rtc/socks.ts]: Seems to me like this function is doing very little more now than "Socks.interpretSocksRequest(byteArray, request);" Do we need both? Also: if an exception is raised, should we not convert it to a promise rejection? 

i.e. Make: `interpretSocksRequest` return the `SocksRequest`, and then just call: `Socks.interpretSocksRequest(new Uint8Array(buffer));` ... what do you think?

**Comment by iislucas** [comment]
nice! Just one suggested tweak to have a think about and then feel free to merge.

File: spec/socks-to-rtc/socks-headers_spec.js (hunk 1 above)
Comment (trevj) [inline spec/socks-to-rtc/socks-headers_spec.js:2]: AFAICT this file already supports everything we need for IPv6 (which is really just decoding IPv6). Some tests for that would be good, of course, so I added some TODOs to the test.

File: src/socks-to-rtc/socks.ts (hunk 2 above)
Comment (trevj) [inline src/socks-to-rtc/socks.ts]: Yes! Inlined.

Regarding the error, it seems that the next thennable in the chain having a rejection argument receives the error - so in this case the handler at line 160 handles this gracefully. Good point!

**Pull request merged**